packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AudioDeviceInfo, VideoDeviceInfo } from '@azure/communication-calling';
+import { CommunicationUserKind } from '@azure/communication-common';
+import { CallState, IncomingCallState, StatefulDeviceManager } from 'calling-stateful-client';
+import {
+  CallAgentOverrides,
+  createMockStatefulCallClient,
+  DeviceManagerOverrides,
+  StatefulCallClientOverrides
+} from '../../../mocks';
+import { AzureCommunicationCallAdapter } from './AzureCommunicationCallAdapter';
+
+const DISPOSE_ERROR_MESSAGE = 'dispose error';
+const GET_CAMERA_ERROR_MESSAGE = 'camera error';
+const GET_MICROPHONE_ERROR_MESSAGE = 'microphone error';
+const GET_SPEAKER_ERROR_MESSAGE = 'speaker error';
+const PERMISSION_ERROR_MESSAGE = 'permission error';
+const JOIN_ERROR_MESSAGE = 'join error';
+const START_ERROR_MESSAGE = 'start error';
+
+const STATEFUL_OVERRIDES: StatefulCallClientOverrides = {
+  getState: () => {
+    return {
+      calls: {},
+      callsEnded: [] as CallState[],
+      incomingCalls: {},
+      incomingCallsEnded: [] as IncomingCallState[],
+      deviceManager: {
+        isSpeakerSelectionAvailable: false,
+        cameras: [] as VideoDeviceInfo[],
+        microphones: [] as AudioDeviceInfo[],
+        speakers: [] as AudioDeviceInfo[],
+        unparentedViews: []
+      },
+      callAgent: undefined,
+      userId: { kind: 'communicationUser', communicationUserId: '' } as CommunicationUserKind
+    };
+  }
+};
+
+const DEVICEMANAGER_EXCEPTION_OVERRIDES: DeviceManagerOverrides = {
+  getCameras: () => {
+    throw new Error(GET_CAMERA_ERROR_MESSAGE);
+  },
+  getMicrophones: () => {
+    throw new Error(GET_MICROPHONE_ERROR_MESSAGE);
+  },
+  getSpeakers: () => {
+    throw new Error(GET_SPEAKER_ERROR_MESSAGE);
+  },
+  askDevicePermission: () => {
+    throw new Error(PERMISSION_ERROR_MESSAGE);
+  }
+};
+
+const CALLAGENT_EXCEPTION_OVERRIDES: CallAgentOverrides = {
+  join: () => {
+    throw new Error(JOIN_ERROR_MESSAGE);
+  },
+  startCall: () => {
+    throw new Error(START_ERROR_MESSAGE);
+  },
+  dispose: () => {
+    return Promise.reject(new Error(DISPOSE_ERROR_MESSAGE));
+  }
+};
+
+describe('AzureCommunicationCallAdapter', () => {
+  test('emits error event when error happens', async () => {
```

Comment:
I'd recommend splitting this into many small tests, each for a single error generated. In that case, I'd also make the mock stateful call client creation local. Something like:

```

const errorString = 'some arbitrary reusable error string';

describe('AzureCommunicationCallAdapter handles errors correctly', () => {
    test('when dispose fails', async () => {
      // Setup
      const statefulCallClient = createMockStatefulCallClient();
     statefulCallClient.dispose = (): void => {
        raise Error(errorString );
     }
     const [adaptor, errorListener] = newCallAdaptorWithErrorListener(statefulCallClient);  // Helper function in this test file to reduce boilerplate.
     // Act
     await adapter.dispose();
     // Assert
     expect(adapter.getState().error?.message).toBe(errorString );
     expect(errorListener.capturedError).toBe(errorString );
 
    }
}
```

Now, this test doesn't have any of the boilerplate that isn't necessary to test *only the `dispose` method*.
And, this test fails iff there's a bug in dispose. You don't have to go hunting through logs on test failure to figure out what's broken.

Reply:
This increases the total number of lines by a bit, but it makes each test a lot smaller and targeted.

Reply:
Thanks, I've updated the test as per these requirements, please take a look if it works for you.